// @truffle/codec/dist/core/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var storage_1 = require("../allocate/storage");
exports.getStorageAllocations = storage_1.getStorageAllocations;
exports.storageSize = storage_1.storageSize;
var abi_1 = require("../allocate/abi");
exports.getAbiAllocations = abi_1.getAbiAllocations;
exports.getCalldataAllocations = abi_1.getCalldataAllocations;
exports.getEventAllocations = abi_1.getEventAllocations;
var memory_1 = require("../allocate/memory");
exports.getMemoryAllocations = memory_1.getMemoryAllocations;
var stack_1 = require("../read/stack");
exports.readStack = stack_1.readStack;
var storage_2 = require("../read/storage");
exports.slotAddress = storage_2.slotAddress;
var decoding_1 = require("./decoding");
exports.decodeVariable = decoding_1.decodeVariable;
exports.decodeEvent = decoding_1.decodeEvent;
exports.decodeCalldata = decoding_1.decodeCalldata;
var abi_2 = require("../encode/abi"); //we have no need to export these at present, but someone might want them?
exports.encodeAbi = abi_2.encodeAbi;
exports.encodeTupleAbi = abi_2.encodeTupleAbi;
